connectors/sources/postgresql.py

```diff
@@ -62,6 +63,231 @@ def all_schemas(self):
+        """
+        try:
+            async with self.engine.connect() as connection:  # pyright: ignore
+                cursor = await connection.execute(text(query))
+                return cursor
+        except Exception as exception:
+            self._logger.warning(
+                f"Something went wrong while executing query. Exception: {exception}"
+            )
+            raise
+
+    async def ping(self):
+        return await anext(
+            self._execute_query(
+                query=self.queries.ping(),
+            )
+        )
+
+    async def get_all_schemas(self):
+        return await anext(self._execute_query(query=self.queries.all_schemas()))
+
+    async def get_tables_to_fetch(self, schema):
+        tables = configured_tables(self.tables)
+        return list(
+            map(
+                lambda table: table[0],  # type: ignore
+                await anext(
+                    self._execute_query(
+                        query=self.queries.all_tables(
+                            database=self.database,
+                            schema=schema,
+                        )
+                    )
+                ),
+            )
+            if is_wildcard(tables)
+            else tables
+        )
+
+    async def get_table_row_count(self, schema, table):
+        [[row_count]] = await anext(
+            self._execute_query(
+                query=self.queries.table_data_count(
+                    schema=schema,
+                    table=table,
+                ),
+            )
+        )
+        return row_count
+
+    async def get_table_primary_key(self, schema, table):
+        return await anext(
+            self._execute_query(
+                query=self.queries.table_primary_key(
+                    schema=schema,
+                    table=table,
+                ),
+            )
+        )
+
+    async def get_table_last_update_time(self, schema, table):
+        return await anext(
+            self._execute_query(
+                query=self.queries.table_last_update_time(
+                    schema=schema,
+                    table=table,
+                ),
+            )
+        )
+
+    async def data_streamer(self, schema, table):
+        """Streaming data from a table
+
+        Args:
+            schema (str): Schema.
+            table (str): Table.
+
+        Raises:
+            exception: Raise an exception after retrieving
+
+        Yields:
+            list: It will first yield the column names, then data in each row
+        """
+        async for data in self._execute_query(
+            query=self.queries.table_data(
+                schema=schema,
+                table=table,
+            ),
+            fetch_many=True,
+            schema=schema,
+            table=table,
+        ):
+            yield data
+
+    async def _execute_query(self, query, fetch_many=False, **kwargs):
+        """Executes a query and yield rows
```

Comment:
I think it's worth to refactor this `fetch_many` flag, `yield_once` logic is also ultra confusing

Reply:
Agree. I can make it two separate functions:
1. `_execute_query`, to simply run the query and yield all.
2. `_execute_query_by_cursor`, which will yield rows one by one. It will also yield column names first if the parameter `include_column_names` is `True` (what `yield_once` does).

I think they were defined in one function, to re-use the retry logic. I actually tried to reuse the `@retryable` annotation, but it doesn't support instance variables:
```python
@retryable(
    retries=self.retry_count, # this is not supported
    interval=0,
    strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
)
```

But I'd like to refactor this in a follow-up PR, to unblock Crest team working on advanced rules.